checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -334,4 +426,147 @@ private static String regexErrorMessage(String s, int expectedGroups, int actual
+    public static List<Integer> getNonNullGroups(String regexp) {
+        int groups = 0;
+        try {
+            Pattern p = Pattern.compile(regexp);
+            groups = getGroupCount(p);
+        } catch (PatternSyntaxException e) {
+            throw new Error(e);
+        }
+        // The list that will hold the groups that are guaranteed to match some part of text that
+        // matches the regex. Initially holds all the groups. The optional groups will be removed.
+        List<Integer> nonNullGroups = new ArrayList<>();
+        for (int i = 1; i <= groups; i++) {
+            nonNullGroups.add(i);
+        }
+
+        // A stack containing indices of the capturing groups that are currently not closed.
+        ArrayDeque<Integer> unclosedCapturingGroups = new ArrayDeque<>();
+
+        // A stack that tracks all open groups (both capturing and non-capturing groups),
+        // indicating whether the group is capturing or not.
+        ArrayDeque<Boolean> isCapturingGroup = new ArrayDeque<>();
+
+        // Index of the most recently opened capturing group.
+        int group = 0;
+
+        // Whenever we encounter any character, we know that it is not escaped. This is because,
+        // when we encountered a '\' which was used to escape something, we traverse till the end
+        // of the escape construct and then only resume normal traversal. This is why we are never
+        // checking if the previous character was a '\'.
+
+        // Optional group here onwards means the ith capturing group, which may not match any part
+        // of a text that matched the regular expression and thus may return null on calls to
+        // matcher.group(i).
+
+        // Special construct means either pure, non-capturing groups that do not capture text and do
+        // not count towards the group total.
+
+        final int length = regexp.length();
+        for (int i = 0; i < length; i++) {
+            // Marks the opening of either a capturing a capturing group or some special
+            // construct.
+            if (regexp.charAt(i) == '(') {
+                boolean isThisCapturingGroup = false;
+                // If '(' is not followed by a '?' or is followed by '?<', it is a capturing group.
+                if ((i < length - 1)
+                        && (regexp.charAt(i + 1) != '?' || regexp.startsWith("?<", i + 1))) {
+                    group += 1;
+                    // Push the group index onto a stack.
+                    unclosedCapturingGroups.push(group);
+                    // This opening bracket marked the opening of a capturing group.
+                    isThisCapturingGroup = true;
+                }
+                // Push the boolean on top of a stack. When we encounter a ')', we will use this
+                // stack to check whether it closes a capturing group or a special construct.
+                isCapturingGroup.push(isThisCapturingGroup);
+                if (isThisCapturingGroup) {
```

Comment:
This is the same condition as the if statement above.  You can merge their bodies.